lms/models/course.py

```diff
@@ -0,0 +1,77 @@
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.ext.mutable import MutableDict
+from sqlalchemy.orm.exc import NoResultFound
+
+from lms.db import BASE
+from lms.models.application_settings import ApplicationSettings
+
+
+class Course(BASE):
+    """An LTI course."""
+
+    __tablename__ = "course"
+
+    #: The authority_provided_id that uniquely identifies the course that these
+    #: settings belong to.
+    authority_provided_id = sa.Column(sa.UnicodeText(), primary_key=True)
+
+    #: The LTI consumer_key (oauth_consumer_key) of the application instance
+    #: that these course settings belong to.
+    consumer_key = sa.Column(
+        sa.String(),
+        sa.ForeignKey("application_instances.consumer_key", ondelete="cascade"),
+        primary_key=True,
+    )
+
+    _settings = sa.Column("settings", MutableDict.as_mutable(JSONB), nullable=False)
+
+    @property
+    def settings(self):
+        """
+        Return this course's settings.
+
+        :rtype: models.ApplicationSettings
+        """
+        return ApplicationSettings(self._settings)
+
+    @classmethod
+    def get(cls, db, authority_provided_id, consumer_key):
+        """Return the course matching `authority_provided_id` and `consumer_key` or None."""
+        try:
+            return cls._get(db, authority_provided_id, consumer_key)
+        except NoResultFound:
+            return None
+
+    @classmethod
+    def insert_if_not_exists(cls, db, authority_provided_id, consumer_key, settings):
+        """
+        Insert the given settings into the DB if none already exist.
+
+        If there's no existing row with the `authority_provided_id` and
+        `consumer_key` then insert the `settings` into the course table with
+        `authority_provided_id` and `consumer_key`.
+
+        If a row matching `authority_provided_id` and `consumer_key` already
+        exists then do nothing.
+        """
+        course = cls(
+            authority_provided_id=authority_provided_id,
+            consumer_key=consumer_key,
+            _settings=settings,
+        )
+
+        try:
+            cls._get(db, authority_provided_id, consumer_key)
+        except NoResultFound:
+            db.add(course)
```

Comment:
The above feels like a service method to me. If we had a service for the getting and setting of courses, you could put the  `record_course` call in there too and manage the transfer of state (from application instance to course) in one place instead of the split here and with the LTI views.

Reply:
`record_course` is specific to a particular context (`LTILaunchResource`) though. I think you could register a service that's only available in that context. I haven't tried it before but I think something like this might work:

```python
config.register_service_factory(
    record_course_service_factory,
    context=LTILaunchResource,
    name="record_course",
)
```

Then I think the views would do something like:

```python
request.find_service(name="record_course").record_course()
```

The point is that any `request.find_service(name="record_course")` outside of an `LTILaunchResource` request would fail. (And in future other `"record_course"` service implementations could potentially be registered for other contexts.)

Reply:
I doubt we'd ever need to register it twice, but I think a "course" service in general would be good.

There a comment on one of the commits:

> The way this is going to work in future is that
> ApplicationInstanceGetter.canvas_sections_enabled() is going to have two
> modes:
> 
> 1. ApplicationInstanceGetter.canvas_sections_enabled(course_settings)
>    will tell you whether sections should be enabled in the context of a
>    given course. This is going to be used to decide whether or not to
>    use sections when launching our app in a course.
> 
> 2. ApplicationInstanceGetter.canvas_sections_enabled() (with no
>    course_settings) argument will just tell you whether sections is
>    enabled for this application instance, without reference to a
>    particular course. This is what we should use to set the course's
>    canvas.sections_enabled setting the first time we see a new course.

I think this would be the natural home for the sections enabled question. Instead of asking the AIGetter to ask the course, ask the course directly (which may internally need to ask the AIGetter... or not)